src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,405 @@
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices
+{
+    [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+    internal static partial class TypeMapLazyDictionary
+    {
+        private ref struct CallbackContext
+        {
+            private LazyExternalTypeDictionary? _externalTypeMap;
+            private LazyProxyTypeDictionary? _proxyTypeMap;
+
+            public LazyExternalTypeDictionary ExternalTypeMap
+            {
+                [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+                get
+                {
+                    _externalTypeMap ??= new LazyExternalTypeDictionary();
+                    return _externalTypeMap;
+                }
+            }
+
+            public LazyProxyTypeDictionary ProxyTypeMap
+            {
+                [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+                get
+                {
+                    _proxyTypeMap ??= new LazyProxyTypeDictionary();
+                    return _proxyTypeMap;
+                }
+            }
+
+            public ExceptionDispatchInfo? CreationException { get; set; }
+        }
+
+        // See assemblynative.hpp for native version.
+        public unsafe struct ProcessAttributesCallbackArg
+        {
+            public void* Utf8String1;
+            public void* Utf8String2;
+            public void* Utf8String3;
+            public int StringLen1;
+            public int StringLen2;
+            public int StringLen3;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypeMapLazyDictionary_ProcessAttributes")]
+        private static unsafe partial void ProcessAttributes(
+            QCallAssembly assembly,
+            QCallTypeHandle groupType,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newExternalTypeEntry,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
+            CallbackContext* context);
+
+        private static TypeName CreateNewTypeName(TypeName typeName, ReadOnlySpan<char> assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = typeName.WithAssemblyName(asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        private static ReadOnlySpan<char> ConvertUtf8ToUtf16(ReadOnlySpan<byte> utf8TypeName)
+        {
+            int defaultUtf8StrLen = 1024;
+            const int MaxUtf8BytesPerChar = 3;
+            int needed = utf8TypeName.Length > (defaultUtf8StrLen * MaxUtf8BytesPerChar)
+                ? Encoding.UTF8.GetCharCount(utf8TypeName)
+                : defaultUtf8StrLen;
+
+            char[] buffer = ArrayPool<char>.Shared.Rent(needed);
```

Comment:
Is this missing matching ArrayPool.Return somewhere?

Reply:
Ugh... yes. I focused too much on the other aspects of its use, apologies.